### PR TITLE
Added aead encrypt decrypt, hash compute compare and raw key agreement

### PR DESCRIPTION
**Pull request by sbailey-arm**
Bit of a big 'un but chunks of it are pretty simple and it saved adding more panics for individual operations.

Signed-off-by: Samuel Bailey <samuel.bailey@arm.com>

### CODE REVIEW

File: src/back/backend_handler.rs [hunk 1]
```diff
@@ -183,9 +183,58 @@ impl BackEndHandler {
...
+            NativeOperation::PsaHashCompare(op_hash_compare) => {
+                let _app_name =
+                    unwrap_or_else_return!(app_name.ok_or(ResponseStatus::NotAuthenticated));
+                let result =
+                    unwrap_or_else_return!(self.provider.psa_hash_compare(op_hash_compare));
+                trace!("psa_aead_decrypt_egress");
+                self.result_to_response(NativeResult::PsaHashCompare(result), header)
+            }
+            NativeOperation::PsaGenerateRandom(_) => {
+                panic!("Unsupported in this PR");
```
Comment (sbailey-arm) [inline src/back/backend_handler.rs:227]: The two `panic!`s are to pass CI. We have pending PRs for `PsaGenerateRandom` and `ListAuthenticators`, so should be immediately removed once those are merged.

File: e2e_tests/tests/per_provider/normal_tests/key_agreement.rs [hunk 2]
```diff
@@ -0,0 +1,78 @@
+// Copyright 2020 Contributors to the Parsec project.
```
Comment (hug-dev) [inline e2e_tests/tests/per_provider/normal_tests/key_agreement.rs:1]: For raw key agreement can you have a test where you generate two keys pairs, do the raw agreement on each private key with the other key pair's public key and verify that the shared secret output is the same?

File: src/providers/mbed_provider/mod.rs [hunk 3]
```diff
@@ -7,8 +7,9 @@ use derivative::Derivative;
 use log::{error, trace};
```
Comment (hug-dev) [inline src/providers/mbed_provider/mod.rs:7]: You might need to add in that file the new supported opcodes. And modify one test that is about it as well I think.

File: src/providers/mbed_provider/mod.rs [hunk 4]
```diff
@@ -225,6 +229,49 @@ impl Provide for MbedProvider {
...
+    ) -> Result<psa_aead_decrypt::Result> {
+        trace!("psa_aead_decrypt ingress");
+        self.psa_aead_decrypt_internal(app_name, op)
+    }
+
+    fn psa_hash_compute(
+        &self,
+        op: psa_hash_compute::Operation,
+    ) -> Result<psa_hash_compute::Result> {
+        trace!("psa_hash_compute");
```
Comment (hug-dev) [inline src/providers/mbed_provider/mod.rs:255]: You forgot the `ingress` there :)

File: Cargo.toml [hunk 5]
```diff
@@ -40,12 +40,15 @@ derivative = "2.1.1"
...
 picky = "5.0.0"
-psa-crypto = { version = "0.3.0" , default-features = false, features = ["operations"], optional = true }
+psa-crypto = { version = "0.4.0" , default-features = false, features = ["operations"], optional = true }
 zeroize = { version = "1.1.0", features = ["zeroize_derive"] }
 picky-asn1-x509 = { version = "0.1.0", optional = true }
 users = "0.10.0"
 libc = "0.2.72"
 
+[patch.crates-io]
+psa-crypto = { git = "https://github.com/parallaxsecond/rust-psa-crypto" }
```
[the remark this answers is not in the document]
Comment (sbailey-arm) [reply Cargo.toml:50]: Yes, latest version with the helper methods has not been released on crates.io.

File: e2e_tests/src/lib.rs [hunk 6]
```diff
@@ -193,6 +194,35 @@ impl TestClient {
         )
     }
 
+    pub fn generate_aes_keys_ccm(&mut self, key_name: String) -> Result<()> {
+        self.generate_key(
+            key_name,
+            Attributes {
+                lifetime: Lifetime::Persistent,
+                key_type: Type::Aes,
+                bits: 0,
```
[the remark this answers is not in the document]
Comment (sbailey-arm) [reply e2e_tests/src/lib.rs:203]: Apparently! Some of the tests were taken from Mbed TLS, which used default (0) bits. I can try changing them but I'm pertty sure in some instances I ran into trouble when changing from 0 to something else.

File: e2e_tests/tests/per_provider/normal_tests/hash.rs [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+#[test]
+fn hash_compute_whirlpool() {
+
+    let mut client = TestClient::new();
+    if !client.is_operation_supported(Opcode::PsaHashCompute) {
+        return;
+    }
+
+    let hash = client.hash_compute(Hash::Sha512, &MESSAGE).unwrap();
```
[the remark this answers is not in the document]
Comment (sbailey-arm) [reply e2e_tests/tests/per_provider/normal_tests/hash.rs:57]: Whoops, I originally used whirlpool but found we don't support it.

File: e2e_tests/tests/per_provider/normal_tests/key_agreement.rs (hunk 2 above)
Comment (sbailey-arm) [reply e2e_tests/tests/per_provider/normal_tests/key_agreement.rs:1]: Good idea 👍

File: Cargo.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (hug-dev) [reply Cargo.toml:50]: Ah! I will release a new version.

File: e2e_tests/tests/per_provider/normal_tests/key_agreement.rs [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    let public_key_1 = client.export_public_key(key_name_1.clone()).unwrap();
+    let public_key_2 = client.export_public_key(key_name_2.clone()).unwrap();
+
+    let shared_secret_1_then_2 = client
+        .raw_key_agreement(RawKeyAgreement::Ecdh, key_name_1.clone(), &public_key_2)
+        .unwrap();
+    let shared_secret_2_then_1 = client
+        .raw_key_agreement(RawKeyAgreement::Ecdh, key_name_2.clone(), &public_key_1)
+        .unwrap();
+    assert_eq!(shared_secret_1_then_2, shared_secret_2_then_1);
```
Comment (hug-dev) [inline e2e_tests/tests/per_provider/normal_tests/key_agreement.rs:120]: So nice to know it actually works 👏

**Review by hug-dev — APPROVED** [review]

**Pull request merged**